dhall/src/Dhall/Schemas.hs

```diff
@@ -0,0 +1,189 @@
+                                file
+                                (Pretty.Text.renderLazy docStream)
+                StandardInput -> do
+                    supportsANSI <- ANSI.hSupportsANSI IO.stdout
+
+                    Pretty.Terminal.renderIO
+                        IO.stdout
+                        (if supportsANSI
+                            then fmap Dhall.Pretty.annToAnsiStyle docStream
+                            else Pretty.unAnnotateS docStream)
+
+        Check -> do
+            if originalText == schemasText
+                then return ()
+                else do
+                    let command = "schemas"
+
+                    let modified = "rewritten"
+
+                    Exception.throwIO CheckFailed{..}
+
+decodeSchema
+    :: Expr Src Void
+    -> Maybe (Map Text (Expr Src Void), Map Text (Expr Src Void))
+decodeSchema
+    (RecordLit [ ("Type", Record _Type), ("default", RecordLit _default) ]) =
+    Just (_Type, _default)
+decodeSchema _ =
+    Nothing
+
+decodeSchemas
+    :: Expr Src Void
+    -> Maybe (Map Text (Map Text (Expr Src Void), Map Text (Expr Src Void)))
+decodeSchemas (RecordLit keyValues) = traverse decodeSchema keyValues
+decodeSchemas  _                    = Nothing
+
+-- | Simplify a Dhall expression using a record of schemas
+simplifyUsingSchemas
+    :: Expr Src Import
+    -- ^ Record of schemas
+    -> Expr Src Import
+    -- ^ Expression to simplify using the supplied schemas
+    -> IO (Expr Src Import)
+simplifyUsingSchemas _schemas expression = do
+    resolvedSchemas    <- Import.load _schemas
+    resolvedExpression <- Import.load expression
+
+    _ <- Core.throws (TypeCheck.typeOf resolvedSchemas)
+    _ <- Core.throws (TypeCheck.typeOf resolvedExpression)
+
+    let normalizedSchemas    = Normalize.normalize resolvedSchemas
+    let normalizedExpression = Normalize.normalize resolvedExpression
+
+    decodedSchemas <- case decodeSchemas normalizedSchemas of
+        Just decodedSchemas -> return decodedSchemas
+        Nothing             -> Exception.throwIO NotASchemaRecord
+
+    let schemasBinding value = Binding{..}
+          where
+            variable   = "schemas"
+            annotation = Nothing
+
+            bindingSrc0 = Nothing
+            bindingSrc1 = Nothing
+            bindingSrc2 = Nothing
+
+    let schemasRewrite subExpression =
+            (Maybe.fromMaybe subExpression . Maybe.listToMaybe) $ do
+                (name, (_Type, _default)) <- Map.toList decodedSchemas
+
+                defaultedRecord <- case subExpression of
+                    RecordLit keyValues -> do
+                        let defaultedKeyValues =
+                                Map.fromList (Map.toList keyValues \\ Map.toList _default)
+
+                        return (RecordLit defaultedKeyValues)
+
+                    _ -> empty
+
+                case TypeCheck.typeOf (Annot (Let (schemasBinding resolvedSchemas) subExpression) (Record _Type)) of
```

Comment:
It seems wasteful to re-check the annotation and the annotated value. Maybe just do the type-equivalence check directly?